README.md
# Luna II

This plasmoid displays the current phase of the moon.

If you click on it, a dialog appears which shows the dates of the important moon 
phases in the same month (last new moon, first quarter, full moon, third quarter
and next new moon).

The format in which these dates are shown can be customized.

You can navigate to previous or future moon phases by clicking the arrow buttons or by 
pressing the arrow keys. Clicking the middle button or pressing the "Home" key 
returns to the current dates.

This is a port to Plasma 5 of:

* Luna QML
* Version 1.4
* <https://store.kde.org/p/1002036/>


package/Changelog
version 2.1.2 (2019-10-13)
- localisation, hr (VladimirMikulic)
version 2.1.1 (2019-09-28)
- localisation, nl (Vistaus)
version 2.1 (2019-09-19)
- Updates by Ismael
  - localisation, es
  - option to make shadow transparent
  - use latitude to set image orientation, replaces the North/South selection
version 2.1-beta3 (2017-07-02)
- got localisation working, fr & ru for starters, need help
- LunaWidget stretches to fit, not fixed size anymore
- replaced the yellow moon disk with one where the colour can be picked
- added a custom.svg image - replace this with whatever image you want
- make calibration marks optional on the config moon image
- the image matches at the quarters now, was out due to assuming constant rate
version 2.1-beta2 (2017-06-29)
- pick lunar image from a list
- current image shown with callibration grid on config page (Tycho & Copernicus marked)
version 2.1-beta (2017-05-07)
- changed the image (edit LunaIcon.qml to use the old one)
- draws the phase shadow on a full moon image instead of an image per day
- maintains circular aspect when resized
version 2.0.3 (2017-03-26)
- stop trampling over theme
- fix tooltips for >=Qt5.5
version 2.0.2 (2017-03-13)
- Issue 3: use lower version of Layouts so it works with >=Qt5.5
version 2.0.1 (2017-02-11)
- zip now now contains the content of the package directory
version 2.0 (2016-12-01)
- modified to be a Plasma 5 plasmoid
version 1.4 (2013-05-16):
- the dialog now scrolls in and out instead of just appearing or disappearing
  when the plasmoid is added in a panel
- split tooltip text in a main text and a subtext
version 1.3 (2012-11-29):
- scrolling the mouse wheel when the mouse cursor is in the popup dialog
  navigates to the previous or next moon phases
- show a more correct description of the current phase in the tooltip
version 1.2 (2012-11-08):
- really make the arrow keys and the "Home" key work again in the popup
  dialog by focusing the dialog each time it is shown
- show correct moon phase in the tooltip
- load the popup dialog only when actually requested (namely when the
  user clicks on the moon image for the first time); this makes the
  plasmoid load faster
version 1.1 (2012-07-18):
- make the arrow keys and the "Home" key work again in the popup dialog
- allow changing the shown phases in the popup dialog with the "Page Up"
  and "Page Down" keys
- allow specifying a custom date format for the dates in the popup
version 1.0 (2012-05-03):
- show the dates in the popup dialog using the user's locale
- allow different formats for the dates in the popup dialog
- add French translation
version 0.4 (2011-12-19):
- use Plasma Components instead of Plasma Widgets for a better layout
  of the popup dialog (KDE Plasma 4.8+ required)
version 0.3 (2011-07-07):
- add tooltip (KDE Plasma 4.7+ required)
- put the dates of the current month's phases in a separate dialog
  (KDE Plasma 4.7+ required)
- the SVG image of the current moon phase is smoother
version 0.2 (2011-06-30):
- add configuration dialog which allows to configure the hemisphere and
  whether the background is drawn
- remove bug: when today is the 30th or 31st of a month, then the dates
  in the months with less days are not shown correctly
- remove bug: the moon image was never updated, now it is updated regularly
version 0.1 (2011-03-24):
- first version


package/contents/code/phases.js
/*
	JavaScript version
	Copyright 2011 Glad Deschrijver <glad.deschrijver@gmail.com>
	with the same license as below.
*/
/*
    This file is part of the kmoon application with explicit permission by the author
    Copyright 1996 Christopher Osburn <chris@speakeasy.org>

    This library is free software; you can redistribute it and/or
    modify it under the terms of the GNU Library General Public
    License as published by the Free Software Foundation; either
    version 2 of the License, or (at your option) any later version.

    This library is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
    Library General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, see <http://www.gnu.org/licenses/>.
*/

.pragma library

/*
** jd.c:
** 1996/02/11
**
** Copyright 1996, Christopher Osburn, Lunar Outreach Services, <chris@speakeasy.org>
** Non-commercial usage license granted to all.
**
** convert a Julian Day number to a struct tm
**
** Parameter:
**   double jd:  Julian day number with fraction of day
**
** Returns:
**   struct tm *event_date:  Date-time group holding year, month, day, hour,
**                and minute of the event
*/

function JDtoDate(jd) // convert a Julian Date to a date-time group
{
	var a, a1, z, b, c, d, e;
	var f, day;
	var event_date = new Date();

	jd += 0.5;
	z = Math.floor(jd);
	f = jd - z;

	if (z < 2299161)
	{
		a = z;
	}
	else
	{
		a1 = Math.floor((z - 1867216.25) / 36524.25);
		a = z + 1 + a1 - Math.floor(a1 / 4);
	}

	b = a + 1524;
	c = Math.floor((b - 122.1) / 365.25);
	d = Math.floor(365.25 * c);
	e = Math.floor((b - d)/30.6001);

	day = b - d - Math.floor(30.6001 * e) + f;

	event_date.setDate(1); // dirty hack, otherwise if today is the 31st of a month and the month below is set to a month with < 31 days, then the month is shifted to the next month, e.g. Feb 31 becomes Mar 2; the correct day is set further below
	if (e < 14)
		event_date.setMonth((e - 1) - 1);
	else
		event_date.setMonth((e - 13) - 1);

	if (event_date.getMonth() > (2 - 1))
		event_date.setFullYear(c - 4716);
	else
		event_date.setFullYear(c - 4715);

	event_date.setDate(Math.floor(day));
	day -= event_date.getDate();
	day *= 24;
	event_date.setHours(Math.floor(day));
	day -= event_date.getHours();
	day *= 60;
	event_date.setMinutes(Math.floor(day));
	day -= event_date.getMinutes();
	day *= 60;
	event_date.setSeconds(Math.floor(day));

	return event_date;
}

function DatetoJD(event_date) // convert a date-time group to a JD with fraction
{
	var y, m;
	var d;
	var a, b;
	var jd;

	y = event_date.getFullYear();
	m = event_date.getMonth() + 1;
	d = event_date.getDate() + (event_date.getHours() / 24.0)
	  + (event_date.getMinutes() / 1440.0)
	  + (event_date.getSeconds() / 86400.0);

	if (m == 1 || m == 2)
	{
		y--;
		m += 12;
	}

	a = Math.floor(y / 100);
	b = 2 - a + Math.floor(a / 4);

	if (y < 1583)
		if ((y < 1582) || (m < 10) || ((m == 10) && (d <= 15)))
			b = 0;

	jd = Math.floor(365.25 * (y + 4716)) + Math.floor(30.6001 * (m+1))
	   + d + b - 1524.5;

	return jd;
}

/*
** misc.h
** 1996/02/11
**
** Copyright 1996, Christopher Osburn, Lunar Outreach Services,
** Non-commercial usage license granted to all.
**
** Miscellaneous routines for moon phase programs
**
*/

function fmod(x, y)
{
	var n = ((x > 0 && y > 0) || (x < 0 && y < 0)) ? Math.floor(x / y) : Math.ceil(x / y);
	return x - n * y;
}

function torad(x) // convert x to radians
{
	x = fmod(x, 360.0); // normalize the angle
	return ((x) * 0.01745329251994329576); // and return the result
}

/*
** moonphase.c
** 1996/02/11
**
** Copyright 1996, Christopher Osburn, Lunar Outreach Services,
** Non-commercial usage license granted to all.
**
** calculate phase of the moon per Meeus Ch. 47
**
** Parameters:
**    int lun:  phase parameter.  This is the number of lunations
**              since the New Moon of 2000 January 6.
**
**    int phi:  another phase parameter, selecting the phase of the
**              moon.  0 = New, 1 = First Qtr, 2 = Full, 3 = Last Qtr
**
** Return:  Apparent JD of the needed phase
*/

function moonphase(k, phi)
{
	var i;                    // iterator to be named later.  Every program needs an i
	var T;                    // time parameter, Julian Centuries since J2000
	var JDE;                  // Julian Ephemeris Day of phase event
	var E;                    // Eccentricity anomaly
	var M;                    // Sun's mean anomaly
	var M1;                   // Moon's mean anomaly
	var F;                    // Moon's argument of latitude
	var O;                    // Moon's longitude of ascending node
	var A = new Array();      // planetary arguments
	var W;                    // added correction for quarter phases

	T = k / 1236.85;                          // (47.3)

	// this is the first approximation.  all else is for style points!
	JDE = 2451550.09765 + (29.530588853 * k)  // (47.1)
	    + T * T * (0.0001337 + T * (-0.000000150 + 0.00000000073 * T));

	// these are correction parameters used below
	E = 1.0                                   // (45.6)
	  + T * (-0.002516 + -0.0000074 * T);
	M = 2.5534 + 29.10535669 * k              // (47.4)
	  + T * T * (-0.0000218 + -0.00000011 * T);
	M1 = 201.5643 + 385.81693528 * k          // (47.5)
	   + T * T * (0.0107438 + T * (0.00001239 + -0.000000058 * T));
	F = 160.7108 + 390.67050274 * k           // (47.6)
	  + T * T * (-0.0016341 * T * (-0.00000227 + 0.000000011 * T));
	O = 124.7746 - 1.56375580 * k             // (47.7)
	  + T * T * (0.0020691 + 0.00000215 * T);

	// planetary arguments
	A[0]  = 0; // unused!
	A[1]  = 299.77 +  0.107408 * k - 0.009173 * T * T;
	A[2]  = 251.88 +  0.016321 * k;
	A[3]  = 251.83 + 26.651886 * k;
	A[4]  = 349.42 + 36.412478 * k;
	A[5]  =  84.66 + 18.206239 * k;
	A[6]  = 141.74 + 53.303771 * k;
	A[7]  = 207.14 +  2.453732 * k;
	A[8]  = 154.84 +  7.306860 * k;
	A[9]  =  34.52 + 27.261239 * k;
	A[10] = 207.19 +  0.121824 * k;
	A[11] = 291.34 +  1.844379 * k;
	A[12] = 161.72 + 24.198154 * k;
	A[13] = 239.56 + 25.513099 * k;
	A[14] = 331.55 +  3.592518 * k;

	// all of the above crap must be made into radians!!!
	// except for E...

	M = torad(M);
	M1 = torad(M1);
	F = torad(F);
	O = torad(O);

	// all those planetary arguments, too!
	for (i = 1; i <= 14; ++i)
		A[i] = torad(A[i]);

	// ok, we have all the parameters, let's apply them to the JDE.
	// (remember the JDE?  this is a program about the JDE...)

	switch(phi)
	{
		// a special case for each different phase.  NOTE!,
		// I'm not treating these in a 0123 order!!!  Pay
		// attention, there, you!

		case 0: // New Moon
			JDE = JDE
			    - 0.40720         * Math.sin(M1)
			    + 0.17241 * E     * Math.sin(M)
			    + 0.01608         * Math.sin(2.0 * M1)
			    + 0.01039         * Math.sin(2.0 * F)
			    + 0.00739 * E     * Math.sin(M1 - M)
			    - 0.00514 * E     * Math.sin(M1 + M)
			    + 0.00208 * E * E * Math.sin(2.0 * M)
			    - 0.00111         * Math.sin(M1 - 2.0 * F)
			    - 0.00057         * Math.sin(M1 + 2.0 * F)
			    + 0.00056 * E     * Math.sin(2.0 * M1 + M)
			    - 0.00042         * Math.sin(3.0 * M1)
			    + 0.00042 * E     * Math.sin(M + 2.0 * F)
			    + 0.00038 * E     * Math.sin(M - 2.0 * F)
			    - 0.00024 * E     * Math.sin(2.0 * M1 - M)
			    - 0.00017         * Math.sin(O)
			    - 0.00007         * Math.sin(M1 + 2.0 * M)
			    + 0.00004         * Math.sin(2.0 * M1 - 2.0 * F)
			    + 0.00004         * Math.sin(3.0 * M)
			    + 0.00003         * Math.sin(M1 + M - 2.0 * F)
			    + 0.00003         * Math.sin(2.0 * M1 + 2.0 * F)
			    - 0.00003         * Math.sin(M1 + M + 2.0 * F)
			    + 0.00003         * Math.sin(M1 - M + 2.0 * F)
			    - 0.00002         * Math.sin(M1 - M - 2.0 * F)
			    - 0.00002         * Math.sin(3.0 * M1 + M)
			    + 0.00002         * Math.sin(4.0 * M1);
			break;

		case 2: // Full Moon
			JDE = JDE
			    - 0.40614         * Math.sin(M1)
			    + 0.17302 * E     * Math.sin(M)
			    + 0.01614         * Math.sin(2.0 * M1)
			    + 0.01043         * Math.sin(2.0 * F)
			    + 0.00734 * E     * Math.sin(M1 - M)
			    - 0.00515 * E     * Math.sin(M1 + M)
			    + 0.00209 * E * E * Math.sin(2.0 * M)
			    - 0.00111         * Math.sin(M1 - 2.0 * F)
			    - 0.00057         * Math.sin(M1 + 2.0 * F)
			    + 0.00056 * E     * Math.sin(2.0 * M1 + M)
			    - 0.00042         * Math.sin(3.0 * M1)
			    + 0.00042 * E     * Math.sin(M + 2.0 * F)
			    + 0.00038 * E     * Math.sin(M - 2.0 * F)
			    - 0.00024 * E     * Math.sin(2.0 * M1 - M)
			    - 0.00017         * Math.sin(O)
			    - 0.00007         * Math.sin(M1 + 2.0 * M)
			    + 0.00004         * Math.sin(2.0 * M1 - 2.0 * F)
			    + 0.00004         * Math.sin(3.0 * M)
			    + 0.00003         * Math.sin(M1 + M - 2.0 * F)
			    + 0.00003         * Math.sin(2.0 * M1 + 2.0 * F)
			    - 0.00003         * Math.sin(M1 + M + 2.0 * F)
			    + 0.00003         * Math.sin(M1 - M + 2.0 * F)
			    - 0.00002         * Math.sin(M1 - M - 2.0 * F)
			    - 0.00002         * Math.sin(3.0 * M1 + M)
			    + 0.00002         * Math.sin(4.0 * M1);
			break;

		case 1: // First Quarter
		case 3: // Last Quarter
			JDE = JDE
			    - 0.62801         * Math.sin(M1)
			    + 0.17172 * E     * Math.sin(M)
			    - 0.01183 * E     * Math.sin(M1 + M)
			    + 0.00862         * Math.sin(2.0 * M1)
			    + 0.00804         * Math.sin(2.0 * F)
			    + 0.00454 * E     * Math.sin(M1 - M)
			    + 0.00204 * E * E * Math.sin(2.0 * M)
			    - 0.00180         * Math.sin(M1 - 2.0 * F)
			    - 0.00070         * Math.sin(M1 + 2.0 * F)
			    - 0.00040         * Math.sin(3.0 * M1)
			    - 0.00034 * E     * Math.sin(2.0 * M1 - M)
			    + 0.00032 * E     * Math.sin(M + 2.0 * F)
			    + 0.00032 * E     * Math.sin(M - 2.0 * F)
			    - 0.00028 * E * E * Math.sin(M1 + 2.0 * M)
			    + 0.00027 * E     * Math.sin(2.0 * M1 + M)
			    - 0.00017         * Math.sin(O)
			    - 0.00005         * Math.sin(M1 - M - 2.0 * F)
			    + 0.00004         * Math.sin(2.0 * M1 + 2.0 * F)
			    - 0.00004         * Math.sin(M1 + M + 2.0 * F)
			    + 0.00004         * Math.sin(M1 - 2.0 * M)
			    + 0.00003         * Math.sin(M1 + M - 2.0 * F)
			    + 0.00003         * Math.sin(3.0 * M)
			    + 0.00002         * Math.sin(2.0 * M1 - 2.0 * F)
			    + 0.00002         * Math.sin(M1 - M + 2.0 * F)
			    - 0.00002         * Math.sin(3.0 * M1 + M);

			W = 0.00306
			  - 0.00038 * E * Math.cos(M)
			  + 0.00026 * Math.cos(M1)
			  - 0.00002 * Math.cos(M1 - M)
			  + 0.00002 * Math.cos(M1 + M)
			  + 0.00002 * Math.cos(2.0 * F);
			if (phi == 3)
				W = -W;
			JDE += W;
			break;

		default: // oops!
			console.log("The Moon has exploded!");
			exit(1);
			break; // unexecuted code
	}

	// now there are some final correction to everything
	JDE = JDE
	    + 0.000325 * Math.sin(A[1])
	    + 0.000165 * Math.sin(A[2])
	    + 0.000164 * Math.sin(A[3])
	    + 0.000126 * Math.sin(A[4])
	    + 0.000110 * Math.sin(A[5])
	    + 0.000062 * Math.sin(A[6])
	    + 0.000060 * Math.sin(A[7])
	    + 0.000056 * Math.sin(A[8])
	    + 0.000047 * Math.sin(A[9])
	    + 0.000042 * Math.sin(A[10])
	    + 0.000040 * Math.sin(A[11])
	    + 0.000037 * Math.sin(A[12])
	    + 0.000035 * Math.sin(A[13])
	    + 0.000023 * Math.sin(A[14]);

	return JDE;
}

//var LUNATION_OFFSET = 953; // original value, I (Glad) don't know why it was chosen as it slows down the startup of the plasmoid
//var LUNATION_OFFSET = -124; // set lunation start in 2010 (faster startup than with the above)
var LUNATION_OFFSET = 0; // a reasonable start lunation is calculated in lunacalc.js

function moonphasebylunation(lun, phi)
{
	var k = lun - LUNATION_OFFSET + phi / 4.0;
	return moonphase(k, phi);
}


package/contents/po/Messages.sh
#!/bin/sh

# based on example found here:
#   https://techbase.kde.org/Development/Tutorials/Localization/i18n_Build_Systems

BUGADDR="https://github.com/wwjjbb/Luna-II/issues"	# MSGID-Bugs

BASEDIR=".."	# root of translatable sources
METADATA="../../metadata.desktop"
WDIR=`pwd`		# working dir

PROJECT=plasma_applet_$(grep "X-KDE-PluginInfo-Name" $METADATA | sed 's/.*=//')
VERSION=$(grep "X-KDE-PluginInfo-Version" $METADATA | sed 's/.*=//')

RCSCRIPT="$WDIR/rc.js"

echo "Preparing rc files"

cd ${BASEDIR}

# additional string for KAboutData
echo 'i18nc("NAME OF TRANSLATORS","Your names");' > $RCSCRIPT
echo 'i18nc("EMAIL OF TRANSLATORS","Your emails");' >> $RCSCRIPT

cd ${WDIR}

echo "Done preparing rc files"

echo "Extracting messages"

cd ${BASEDIR}

# see above on sorting
find . -name '*.qml' -o -name '*.js' | sort > ${WDIR}/infiles.list

cd ${WDIR}
xgettext --from-code=UTF-8 -C -kde -ci18n -ki18n:1 -ki18nc:1c,2 -ki18np:1,2 -ki18ncp:1c,2,3 -ktr2i18n:1 \
	-kI18N_NOOP:1 -kI18N_NOOP2:1c,2 -kaliasLocale -kki18n:1 -kki18nc:1c,2 -kki18np:1,2 -kki18ncp:1c,2,3 \
	--msgid-bugs-address="${BUGADDR}" \
	--files-from=infiles.list -D ${BASEDIR} -D ${WDIR} -o ${PROJECT}.pot || { echo "error while calling xgettext. aborting."; exit 1; }

echo "Done extracting messages"


echo "Merging translations"

catalogs=$(find . -name '*.po')
for cat in $catalogs; do
  echo $cat
  msgmerge -o $cat.new $cat ${PROJECT}.pot
  mv $cat.new $cat
done

echo "Done merging translations"

echo "Generating mo files"

catalogs=$(find . -name '*.po')
for cat in $catalogs; do
  echo $cat
	catdir=locale/${cat%.*}/LC_MESSAGES
	mkdir -p $catdir
  msgfmt $cat -o $catdir/$PROJECT.mo
done

echo "Done generating mo files"



echo "Cleaning up"
cd ${WDIR}
rm infiles.list
rm $RCSCRIPT

echo "Done"


package/contents/code/shadowcalcs.js
/**

		Copyright 2017 Bill Binder <dxtwjb@gmail.com>

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, see <http://www.gnu.org/licenses/>.

*/

.pragma library

var rlookup;
var radius;

function sqr(x) {
  return x * x;
}

function setup(r) {
    radius = r;
    rlookup = [];
    for (var z=-r; z<=r; z++) {
        rlookup[z] = r*Math.sqrt(1 - sqr(z*1.0/r));
    }
    //console.log("Radius: " + r.toString());
    //console.log("Size:   " + rlookup.length.toString());

    return radius;
}

function get(z) {
    return rlookup[Math.abs(z)];
}


package/contents/code/lunacalc.js
/**

    Copyright 1998, 2000  Stephan Kulow <coolo@kde.org>
    Copyright 2008 by Davide Bettio <davide.bettio@kdemail.net>
    Copyright (C) 2009, 2011, 2012, 2013 Glad Deschrijver <glad.deschrijver@gmail.com>

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, see <http://www.gnu.org/licenses/>.

*/

var lunation = 0;

function getLunation(time)
{
//	var lunation = 0;
	var nextNew = new Date(0);

	// obtain reasonable start value for lunation so that the while loop below has a minimal amount of iterations (for faster startup of the plasmoid)
	var reference = 947178885000; // number of milliseconds between 1970-01-01 00:00:00 and 2000-01-06 18:14:45 (first new moon of 2000, see lunation in phases.js)
	var lunationDuration = 2551442877; // number of milliseconds in 29.530588853 days (average lunation duration, same number as in (47.1) in phases.js)
//	var lunationDuration = 2583360000; // number of milliseconds in 29.9 days (maximum lunation duration, see wikipedia on synodic month); use this if the bug ever appears that the lunar phases displayed at startup are too much in the future
	var lunation = Math.floor((time.getTime() - reference) / lunationDuration);

	do {
		var JDE = Phases.moonphasebylunation(lunation, 0);
		nextNew = Phases.JDtoDate(JDE);
		lunation++;
	} while (nextNew < time);

	lunation -= 2;
	return lunation;
}

function getPhasesByLunation(lunation)
{
	var phases = new Array();
	phases[0] = Phases.JDtoDate(Phases.moonphasebylunation(lunation, 0)); // new moon
	phases[1] = Phases.JDtoDate(Phases.moonphasebylunation(lunation, 1)); // first quarter
	phases[2] = Phases.JDtoDate(Phases.moonphasebylunation(lunation, 2)); // full moon
	phases[3] = Phases.JDtoDate(Phases.moonphasebylunation(lunation, 3)); // last quarter
	phases[4] = Phases.JDtoDate(Phases.moonphasebylunation(lunation+1, 0)); // next new moon
	return phases;
}

function getTodayPhases()
{
	var today = new Date();
	lunation = getLunation(today);
	return getPhasesByLunation(lunation);
}

function getPreviousPhases()
{
	lunation--;
	return getPhasesByLunation(lunation);
}

function getNextPhases()
{
	lunation++;
	return getPhasesByLunation(lunation);
}

function reloadPhases()
{
	return getPhasesByLunation(lunation);
}

function getImageNumber(quarter,fraction)
{
	var mapping = new Array();
	mapping[0] = {first:0, last:7 }
	mapping[1] = {first:7, last:13 }
	mapping[2] = {first:13, last:21 }
	mapping[3] = {first:21, last:28 }

    var range = mapping[quarter].last - mapping[quarter].first;
	return Math.round(range * fraction + mapping[quarter].first);
}

function getCurrentPhase() // this function assumes that today is between phases[0] (last new moon) and phases[4] (next new moon)
{
	var oneDay = 1000 * 60 * 60 * 24;
	var today = new Date().getTime();
	var phases = getTodayPhases();

	// Find which quarter is the current one
	var qnum = 0;
	while (today > phases[qnum+1] && qnum < 3) {
	   qnum++;
	}

	// Work out an image number in the range 0-28. This interpolates
	// between the times of the quarters, allowing for the quarters to
	// have different durations. The result is going to be similar to
	// the day number worked out below, but should match the sky better.
	var quarterTime = phases[qnum+1].getTime() - phases[qnum].getTime();
	var sinceQuarter = today - phases[qnum].getTime();
	var imgNumber = getImageNumber(qnum,sinceQuarter/quarterTime);

	// set time for all phases to 00:00:00 in order to obtain the correct phase for today (these changes should be local)
	for (var i = 0; i < 5; i++) {
		phases[i].setHours(0);
		phases[i].setMinutes(0);
		phases[i].setSeconds(0);
	}

	// if today <= first quarter, calculate day since last new moon
	var daysFromFirstQuarter = Math.floor((today - phases[1].getTime()) / oneDay);
	if (daysFromFirstQuarter == 0)
		return {number: 7, text: i18n("First Quarter"), subText: "", img: imgNumber};
	else if (daysFromFirstQuarter < 0) {
		var daysFromLastNew = Math.floor((today - phases[0].getTime()) / oneDay);
		if (daysFromLastNew == 0)
			return {number: 0, text: i18n("New Moon"), subText: "", img: imgNumber};
		else if (daysFromLastNew == 1)
			return {number: 1, text: i18n("Waxing Crescent"), subText: i18n("Yesterday was New Moon"), img: imgNumber};
		else // assume that today >= last new moon
			return {number: daysFromLastNew, text: i18n("Waxing Crescent"), subText: i18n("%1 days since New Moon", daysFromLastNew), img: imgNumber};
	}

	// if today >= third quarter, calculate day until next new moon
	var daysFromThirdQuarter = Math.floor((today - phases[3].getTime()) / oneDay);
	if (daysFromThirdQuarter == 0)
		return {number: 21, text: i18n("Last Quarter"), subText: "", img: imgNumber};
	else if (daysFromThirdQuarter > 0) {
		var daysToNextNew = -Math.floor((today - phases[4].getTime()) / oneDay);
		if (daysToNextNew == 0)
			return {number: 0, text: i18n("New Moon"), subText: "", img: imgNumber};
		else if (daysToNextNew == 1)
			return {number: 27, text: i18n("Waning Crescent"), subText: i18n("Tomorrow is New Moon"), img: imgNumber};
		else // assume that today <= next new moon
			return {number: 28 - daysToNextNew, text: i18n("Waning Crescent"), subText: i18n("%1 days to New Moon", daysToNextNew), img: imgNumber};
	}

	// in all other cases, calculate day from or until full moon
	var daysFromFullMoon = Math.floor((today - phases[2].getTime()) / oneDay);
	if (daysFromFullMoon == 0)
		return {number: 14, text: i18n("Full Moon"), subText: "", img: imgNumber};
	else if (daysFromFullMoon == -1)
		return {number: 13, text: i18n("Waxing Gibbous"), subText: i18n("Tomorrow is Full Moon"), img: imgNumber};
	else if (daysFromFullMoon < -1)
		return {number: 14 + daysFromFullMoon, text: i18n("Waxing Gibbous"), subText: i18n("%1 days to Full Moon", -daysFromFullMoon), img: imgNumber};
	else if (daysFromFullMoon == 1)
		return {number: 15, text: i18n("Waning Gibbous"), subText: i18n("Yesterday was Full Moon"), img: imgNumber};
	else if (daysFromFullMoon > 1)
		return {number: 14 + daysFromFullMoon, text: i18n("Waning Gibbous"), subText: i18n("%1 days since Full Moon", daysFromFullMoon), img: imgNumber};

	// this should never happen:
	console.log("We cannot count :-(");
	return {number: -1, text: ""};
}
